comm/src/main/scala/coop/rchain/comm/protocol.scala

```diff
@@ -0,0 +1,163 @@
+package coop.rchain.comm
+
+import coop.rchain.kademlia
+import coop.rchain.comm.protocol.routing._
+import scala.util.{Failure, Success, Try}
+import scala.concurrent.duration.{Duration, MILLISECONDS}
+
+/**
+  * Implements broadcasting and round-trip (request-response) messaging
+  * for higher level protocols.
+  */
+trait ProtocolHandler {
+  /**
+    * The node that anchors this handler; `local` becomes the source
+    * for outgoing communications.
+    */
+  def local: ProtocolNode
+
+  /**
+    * Send a message to a single, remote node, and wait up to the
+    * specified duration for a response.
+    */
+  def roundTrip(msg: ProtocolMessage,
+                remote: ProtocolNode,
+                timeout: Duration = Duration(500, MILLISECONDS)): Try[ProtocolMessage]
+
+  /**
+    * Asynchronously broadcast a message to all known peers.
+    */
+  def broadcast(msg: ProtocolMessage): Seq[Try[Unit]]
+}
+
+/**
+  * A `PeerNode` that knows how to send and receive messages. The
+  * `ping` method for Kademlia is here.
+  */
+class ProtocolNode(id: NodeIdentifier, endpoint: Endpoint, handler: ProtocolHandler)
+    extends PeerNode(id, endpoint)
+    with kademlia.Peer {
+
+  def this(peer: PeerNode, handler: ProtocolHandler) =
+    this(peer.id, peer.endpoint, handler)
+
+  private var _seq = 0L
+  def seq: Long = _seq synchronized {
+    _seq += 1
+    _seq
+  }
+
+  override def ping: Try[Duration] =
+    ProtocolMessage.ping(handler.local) match {
+      case Some(ping) =>
+        handler.roundTrip(PingMessage(ping, System.currentTimeMillis), this) match {
+          case Success(pong) =>
+            ping.header match {
+              case Some(incoming) =>
+                Success(Duration(pong.timestamp - incoming.timestamp, MILLISECONDS))
+              case _ => Failure(new Exception("ping failed"))
+            }
+          case Failure(ex) => Failure(ex)
+        }
+      case None => Failure(new Exception("ping failed"))
+    }
+}
+
+/**
+  * `ProtocolMessage` insulates protocol handlers from protocol buffer
+  * clutter.
+  */
+trait ProtocolMessage {
+  val proto: Protocol
+  val timestamp: Long
+}
+
+case class PingMessage(proto: Protocol, timestamp: Long) extends ProtocolMessage
+case class PongMessage(proto: Protocol, timestamp: Long) extends ProtocolMessage
+
+/**
+  * Utility functions for working with protocol buffers.
+  */
+object ProtocolMessage {
+
+  implicit def toProtocolBytes(x: String) =
+    com.google.protobuf.ByteString.copyFromUtf8(x)
+  implicit def toProtocolBytes(x: Array[Byte]) =
+    com.google.protobuf.ByteString.copyFrom(x)
+  implicit def toProtocolBytes(x: Seq[Byte]) =
+    com.google.protobuf.ByteString.copyFrom(x.toArray)
+
+  def toPeer(h: Option[Header]): Option[PeerNode] =
+    for {
+      h <- h
```

Comment:
What does "h <- h" accomplish?  I have a basic understanding of for comprehensions but I don't see why this works.  Why does the first execution of h <- h not overwrite the h on the right hand side?

Reply:
The `for` introduces a scope (I don't have all the Scala lingo down). You're right that it's weird looking, and I should probably change one of the `h`s to something like `header` to make it maybe clearer.

Reply:
It does more or less the same as
```
  def toPeer(h: Option[Header]): Option[PeerNode] =
    h.flatMap { h =>
        h.sender
      }
      .flatMap { node =>
        Some(
          PeerNode(NodeIdentifier(node.id.toByteArray),
                   Endpoint(node.host.toStringUtf8, node.tcpPort, node.udpPort)))
      }
```

Reply:
Which I guess is the same as
```
  def toPeer(h: Option[Header]): Option[PeerNode] =
    h.flatMap { h =>
        h.sender
      }
  .map(node => PeerNode(NodeIdentifier(node.id.toByteArray),
                        Endpoint(node.host.toStringUtf8, node.tcpPort, node.udpPort)))
```
in this case.